### ISSUE TITLE
Timers for Main Scenario Roulette Cinematics

### ISSUE THREAD

**Issue by Marakalaid** [quisquous/cactbot#1046]
## Summary

It would be very useful to have accurate countdowns for the many cinematics of Castrum Meridianum and the Praetorium, similar to how Cactbot provides timers for boss phase transitions.

Would that be possible?

![Mock](https://user-images.githubusercontent.com/61011555/74435330-1ab13400-4e5c-11ea-9aa3-03a503961e71.jpg)

**Comment by quisquous** [comment]
Feel free to submit timelines for this content! Patches are very welcome.

**Comment by Jaehyuk-Lee** [comment]
Whooo... this is good

**Comment by Marakalaid** [comment]
> 
> 
> Feel free to submit timelines for this content! Patches are very welcome.

I'm happy to put in the work needed, though this is my first time doing this. Is there any documentation for making your own timelines for Cactbot? I've looked at files for other duties but it's mostly greek to me.

For now I've tracked the duration of each cinematic, as well as the first message that appears in the log for each one (I assume that's the best way to track when it happens). But I don't know how to turn all this info into something useable in Cactbot.

**Comment by Akurosia** [comment]
Try have a look into the timeline guide: https://github.com/quisquous/cactbot/blob/master/docs/TimelineGuide.md

**Comment by Jaehyuk-Lee** [comment]
btw, this "cinematic" timeline should be very long, so I think this needs exception for not filtered by "ShowTimerBarsAtSeconds" setting.

**Comment by quisquous** [comment]
You could give it a duration as a workaround to not changing the window to see future bars.

**Comment by 1unie** [comment]
From experience, just a note that cutscenes might have different lengths depending on both game language and audio language. I don't think it really matters since an estimate is better than nothing but just wanted to flag in case.

**Comment by Jaehyuk-Lee** [comment]
ah, yes, that's right. It can be different.

**Comment by bbeenn1** [comment]
Since the Moogle Event for the Tomestones of Law is open until 5.3 drops, and more people may be running these I am bumping this and asking if anyone has already done the timelines in Cactbot. (I only recently started using it and have not had opportunity to make my own timelines).

I have a _rough_ scene length overview for The Praetorium but have not had the chance to find specific (raw) code for when the scenes start so I had built triggers at one point (in Triggernometry) based on the first NPC dialogue in a scene. This just has a text aura start when you enter and adjusts a variable when each scene starts, clears the variable when the countdown <= 0 or you enter combat, and toggles the visibility of the aura in between.
I can make it a point to export logs from my next run if it would help.

This is what it looks like in Triggernometry:
![image](https://user-images.githubusercontent.com/5580320/83178202-104a3b80-a0d5-11ea-8656-4a2b5ad26d56.png)

The export of that folder:
[Triggernometry_Praetorium CS Tasa.zip](https://github.com/quisquous/cactbot/files/4697474/Triggernometry_Praetorium.CS.Tasa.zip)

**Comment by quisquous** [comment]
Patches still welcome.

**Comment by miaonima** [comment]
I found that the game emits "01:Changed Zone to" when the player is about to regain control of the character, so I think it is possible to write a trigger against this message. Here is a demo that makes a bell sound to notify the player when the cutscene is over:

```
    {
        zoneId: ZoneId.ThePraetorium,
        triggers: [
            {
                id: "Praetorium Cutscene Over Alert",
                regex: /01:Changed Zone to/,
                sound: 'Long',
            }
        ]
    }

```

However, this also get triggered when the play enter the mission or reconnect after force disconnecting/closing the game via task manager during the cutscenes.

Is there any way I can write this in a standlone js file instead of user/raidboss.js, just as the timeline or js files shipped with cactbot? Thanks and feel free to comment.

**Comment by MaikoTan** [comment]
`user/raidboss/any-name-you-like.js`

see https://github.com/quisquous/cactbot/blob/main/docs/CactbotCustomization.md#user-directory-overview

**Comment by miaonima** [comment]
Thanks for your helpful suggestions! Currently I am trying to suppress this trigger before the first cutscene but the data variable trick does not seem to work. I think data variable gets reset right before a zone change. Here is what I have so far:

```
{
    zoneId: ZoneId.ThePraetorium,
    triggers: [
        {
            id: "Praetorium Cutscene Over Alert",
            netRegex: NetRegexes.changeZone({ id: ZoneId.MatchAll }),
            // TODO: Need to suppress the first occurrence. Following does not work.
            //preRun: function(data, matches) {
            //    data.changeZoneCount = (data.changeZoneCount || 0) + 1;
            //},
            //condition: function(data, matches) {
            //    return data.changeZoneCount >= 2;
            //},
            infoText: {
                cn: '可行动',
                en: 'Player Ready',
            },
            sound: 'Long',
        }
    ]
}
```

I tried this a few times but data.changeZoneCount is always 1. Any workaround? Suggestions welcome.

**Comment by quisquous** [comment]
I think it would be better to make a real timeline for the main scenario instances, and then we can add timeline triggers.

**Comment by MaikoTan** [comment]
Yeah, in fact I am creating such a timeline, but it is hard to determine how long the cutscene would playing, just like mentioned above, it depends on the player's game language and/or audio language.

**Comment by quisquous** [comment]
I think if you can figure it out from the (parser) language, then timeline injection seems reasonable to me.  I didn't realize the cutscenes had variable lengths.  What a mess!

**Comment by bbeenn1** [comment]
Yep, Different lengths depending on language, and there are some of them that are different by up to 30 seconds or so in the MSQ Roulette.
The triggers I had made in English are triggered by the first sentence of each scene by the NPCs in the scene - the only downside for me is that some of the scenes take almost 30 seconds to get through the beginning of the dialogue so it can be parsed. 
My Triggers stopped popping up on screen at some point, but that is more my own meddling than them not being set - the time values are still set when the scenes start.
